addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/MiHomeSubdevicesHandler.java

```diff
@@ -0,0 +1,651 @@
+
+            boolean isNewDeviceFound = type.equals(deviceType) && gateway == gatewayID && !oldIDs.contains(deviceID);
+            if (isNewDeviceFound) {
+                return deviceObj;
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        ThingStatus status = getThing().getStatus();
+        if (status == ThingStatus.ONLINE) {
+            JsonObject deviceData = gatewayHandler.getSubdeviceData(mihomeID);
+            if (deviceData != null) {
+                if (command instanceof RefreshType && channelUID.getId().equals(CHANNEL_STATE)) {
+                    updateThingState(deviceData, channelUID);
+                } else {
+                    logger.warn("Unsupported command {} for channel with UID {}", command.toFullString(),
+                            channelUID.getAsString());
+                }
+            } else {
+                logger.warn("Cann't execute command {}. No data for device with ID {}", command, this.mihomeID);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Subdevice " + mihomeID + " is missing, it might be deleted!");
+            }
+        } else {
+            logger.warn("Device with ID {} is in status {}. Not able to handle command {} for channel with UID {}",
+                    this.mihomeID, status, command.toFullString(), channelUID.getAsString());
+        }
+
+    }
+
+    private void updateThingState(JsonObject deviceData, ChannelUID channelUID) {
+        ThingTypeUID thingTypeUID = getThing().getThingTypeUID();
+
+        String channelID = channelUID.getId();
+        State state = null;
+
+        switch (thingTypeUID.getId()) {
+            case MiHomeBindingConstants.THING_ID_MOTION_SENSOR:
+                state = getMotionSensorState(deviceData, channelID);
+                break;
+            case MiHomeBindingConstants.THING_ID_OPEN_SENSOR:
+                state = getOpenSensorState(deviceData, channelID);
+                break;
+            case MiHomeBindingConstants.THING_ID_ENERGY_MONITOR:
+                state = getEnergyMonitorState(deviceData, channelID);
+                break;
+            default:
+                logger.warn("Channel with UID {} won't be updated! It is not supported from thing type {}",
+                        channelUID.getAsString(), thingTypeUID.getAsString());
+        }
+
+        if (state != null) {
+            logger.debug("About to update state for channel {} to {}", channelUID.getAsString(), state.toFullString());
+            updateState(channelUID, state);
+        } else {
+            logger.info("Channel with UID {} won't be updated. Server response is incomplete.",
+                    channelUID.getAsString());
+        }
+    }
+
+    private State getEnergyMonitorState(JsonObject deviceData, String channelID) {
+        State state = null;
```

Comment:
Not in this case, please note that if this method returns null, I won't call updateState() at all.